src/backend/gl/src/command.rs

```diff
@@ -141,33 +153,158 @@ pub const RESET: [Command; 14] = [
 struct Cache {
     primitive: gl::types::GLenum,
     index_type: c::IndexType,
+    current_vbs: Option<c::pso::VertexBufferSet<Resources>>,
     attributes: [Option<BufferElement>; c::MAX_VERTEX_ATTRIBUTES],
     resource_binds: [Option<gl::types::GLenum>; c::MAX_RESOURCE_VIEWS],
     scissor: bool,
     target_dim: (u16, u16, u16),
     stencil: Option<s::Stencil>,
-    //blend: Option<s::Blend>,
+    // blend: Option<s::Blend>,
     cull_face: s::CullFace,
     draw_mask: u32,
+
+    program: Program,
+    constant_buffer: Option<c::pso::ConstantBufferParam<Resources>>,
+    resource_view: Option<c::pso::ResourceViewParam<Resources>>,
+    scissor_test: Option<Rect>,
+    depth_state: Option<s::Depth>,
+    blend_state: Option<(c::ColorSlot, s::Color)>,
+    blend_color: Option<ColorValue>,
+    viewport: Option<Rect>,
+    rasterizer: Option<s::Rasterizer>,
+    framebuffer: Option<(Access, FrameBuffer)>,
+    index: Buffer,
 }
 
 impl Cache {
     pub fn new() -> Cache {
         Cache {
             primitive: 0,
             index_type: c::IndexType::U16,
+            current_vbs: None,
             attributes: [None; c::MAX_VERTEX_ATTRIBUTES],
             resource_binds: [None; c::MAX_RESOURCE_VIEWS],
             scissor: false,
             target_dim: (0, 0, 0),
             stencil: None,
             cull_face: s::CullFace::Nothing,
-            //blend: None,
+            // blend: None,
             draw_mask: 0,
+
+            program: 0,
+            constant_buffer: None,
+            resource_view: None,
+            scissor_test: None,
+            depth_state: None,
+            blend_state: None,
+            blend_color: None,
+            viewport: None,
+            rasterizer: None,
+            framebuffer: None,
+            index: 0,
+        }
+    }
+
+    fn bind_program(&mut self, program: Program) -> Option<Command> {
+        if program == self.program {
```

Comment:
nit: I think a simple `if/else` for the case of small branches should be preferred over `return`